exotica_core/include/exotica_core/visualization_meshcat_types.h

```diff
@@ -0,0 +1,447 @@
+    {
+        file_name = ParsePath(file_name_in);
+        if (format_in == "")
+        {
+            std::string::size_type pos = file_name.rfind(".");
+            if (pos != std::string::npos)
+            {
+                format = file_name.substr(pos + 1);
+            }
+        }
+        std::transform(format.begin(), format.end(), format.begin(), ::tolower);
+        url = ParsePath(url_in);
+        std::ifstream file(file_name, std::ios::binary | std::ios::ate);
+        std::streamsize size = file.tellg();
+        file.seekg(0, std::ios::beg);
+
+        buffer.resize(size);
+        file.read(buffer.data(), size);
+
+        data.ptr = buffer.data();
+        data.size = buffer.size();
+    };
+
+    GeometryMesh(const GeometryMesh& other) : Geometry(other.type, other.uuid)
+    {
+        buffer = other.buffer;
+        url = other.url;
+        resources = other.resources;
+        matrix = other.matrix;
+        format = other.format;
+        data.ptr = buffer.data();
+        data.size = buffer.size();
+    };
+    std::vector<char> buffer;
+    std::string file_name;
+    std::string format;
+    msgpack::v1::type::raw_ref data;
+    std::string url;
+    std::map<std::string, std::string> resources;
+    std::vector<double> matrix = {1.0, 0.0, 0.0, 0.0, 0.0, 1.0, 0.0, 0.0, 0.0, 0.0, 1.0, 0.0, 0.0, 0.0, 0.0, 1.0};
+    MSGPACK_DEFINE(uuid, type, format, resources, url, data, matrix)
+};
+
+struct ArrayFloat
+{
+    ArrayFloat() = default;
+
+    ArrayFloat(double* data_in, unsigned int size)
+    {
+        data.resize(size);
+        for (unsigned int i=0; i < size; ++i)
+            data[i] = data_in[i];
+        array = msgpack::v1::type::ext(0x17, reinterpret_cast<const char*>(data.data()), sizeof(float) * data.size());
+    }
+
+    ArrayFloat(const ArrayFloat& other)
+    {
+        itemSize = other.itemSize;
+        normalized = other.normalized;
+        data = other.data;
+        array = msgpack::v1::type::ext(0x17, reinterpret_cast<const char*>(data.data()), sizeof(float) * data.size());
+    }
+
+    int itemSize = 3;
+    std::string type = "Float32Array";
+    bool normalized = false;
+    std::vector<float> data;
+    msgpack::v1::type::ext array;
+
+    MSGPACK_DEFINE(itemSize, type, normalized, array)
+};
+
+struct ArrayInt
+{
+    ArrayInt() = default;
+
+    ArrayInt(unsigned int* data, unsigned int size)
+    {
+        array.resize(size);
+        for (unsigned int i=0; i < size; ++i)
+            array[i] = data[i];
+    }
+
+    int itemSize = 3;
+    std::string type = "Uint32Array";
+    bool normalized = false;
+    std::vector<uint32_t> array;
+
+    MSGPACK_DEFINE(itemSize, type, normalized, array)
+};
+
+struct GeometryMeshBufferData
+{
+    GeometryMeshBufferData() = default;
+    GeometryMeshBufferData(shapes::ShapePtr shape_in)
+    {
+        std::shared_ptr<shapes::Mesh> shape = std::static_pointer_cast<shapes::Mesh>(shape_in);
```

Comment:
Why do you need to cast this to a shared ptr? And why cant you just take the reference of the pointer or the const ptr?